### PR TITLE
Annotation-less P2P shuffling

### LINKED ISSUE #7716: [Tracking] Annotations-related issues with P2P
The implementation of P2P shuffling/rechunking relies on annotations. Unfortunately, there have been a number of issues caused by annotations getting dropped before the execution of the P2P shuffle/rechunk.

* https://github.com/dask/dask/issues/10118
* https://github.com/dask/dask/issues/10074
* #7715
* `da.rechunk` only works if `"optimization.fuse.active": False` is configured due to dask/dask#7036
* https://github.com/dask/dask/pull/10123

**Additional thoughts**

To get out of the mess we're currently in, I'm wondering if making annotations available from within tasks as suggested in https://github.com/dask/distributed/issues/6932 would make it easier to test that they do not get lost.

Together with the work on High-Level Expressions (dask/dask#7933), we should rethink the current API for annotations. It seems to be too easy to drop annotations somewhere along the way.

### PR DESCRIPTION
**Pull request by hendrikmakait**
Closes #7716 by avoiding annotations altogether.

**Core idea:**
* Instead of propagating information through annotations, we optimistically schedule output tasks on any worker, identify the correct worker and reschedule onto that worker.

**Assumption:**
* Rescheduling each task at most once will not add significant overhead.

**Open questions:**

* How do we deal with pre-existing worker restrictions that do not match the output worker we have chosen for the given partition?

- [x] Tests added / passed
- [x] Passes `pre-commit run --all-files`

### CODE REVIEW

**Comment by mrocklin** [comment]
Cc @rjzamora and @d-v-b so that they're aware of this possibility

**Comment by hendrikmakait** [comment]
Here are results from an A/B test on this, it looks like there is no significant performance impact: 
Screenshot 2023-04-27 at 16 54 24

File: distributed/shuffle/_merge.py
Comment (fjetter) [inline distributed/shuffle/_merge.py:218]: I suggest to still keep the annotations. They will not hurt and we might want to utilize them *iff* they are available.

File: distributed/shuffle/_scheduler_extension.py
Comment (fjetter) [inline distributed/shuffle/_scheduler_extension.py:132]: I think we should also check if the shuffle this run_id points to is still valid. We don't want to have stale requests modify the restrictions.

File: distributed/shuffle/_scheduler_extension.py
Comment (fjetter) [inline distributed/shuffle/_scheduler_extension.py:186]: why can't this logic be the same as before when it comes to existing worker restrictions?

**Comment by fjetter** [comment]
I think this all makes sense. I wonder if we still want to keep the annotations path around for the "happy case" where HLGs and annotations just work

File: distributed/shuffle/tests/test_shuffle.py
Comment (wence-) [inline distributed/shuffle/tests/test_shuffle.py:136]: This test doesn't make sense any more after these changes right? So xfail is wrong and it should be deleted?

File: distributed/shuffle/tests/test_rechunk.py
Comment (wence-) [inline distributed/shuffle/tests/test_rechunk.py:164]: This is also now no longer expected to raise an error and so the `xfail` is a bit of a misnomer I think?

File: distributed/shuffle/_scheduler_extension.py
Comment (wence-) [inline distributed/shuffle/_scheduler_extension.py:132]: Add a helpful error message here?

**Review by wence- — COMMENTED** [review]
A few minor queries.

File: distributed/shuffle/tests/test_shuffle.py
Comment (hendrikmakait) [reply distributed/shuffle/tests/test_shuffle.py]: Yup, this is still work in progress; I've adjusted the test.

File: distributed/shuffle/tests/test_rechunk.py
Comment (hendrikmakait) [reply distributed/shuffle/tests/test_rechunk.py]: Yup, this is still work in progress; I've adjusted the test.

File: distributed/shuffle/_scheduler_extension.py
Comment (hendrikmakait) [reply distributed/shuffle/_scheduler_extension.py]: I've adjusted the logic. The only unhandled situation I can imagine is if a task is fused with a follow-up task, which has worker restrictions of its own. In this case, we may ignore those. A fix for this behavior would be adding a shuffle-internal mechanism for transferring the output partition to another worker.

File: distributed/shuffle/_scheduler_extension.py
Comment (hendrikmakait) [reply distributed/shuffle/_scheduler_extension.py:133]: We should probably document this somewhere, but if it's in `ShuffleSchedulerExtension.states`, it's an active and valid shuffle instance. Otherwise, it would have been dropped from there.

**Comment by hendrikmakait** [comment]
> How do we deal with pre-existing worker restrictions that do not match the output worker we have chosen for the given partition?

If the worker restrictions are applied to the shuffle, they apply to all tasks. Thus, we can use the worker restrictions of the barrier task to bootstrap our set of valid workers. 

As things stand right now, I cannot think of another scenario that would add worker restrictions to tasks. With task fusion, annotations are still getting lost (dask/dask#7036). If this changes, we need to revisit this.

File: distributed/shuffle/_scheduler_extension.py
Comment (hendrikmakait) [inline distributed/shuffle/_scheduler_extension.py:184]: XREF: #7816

File: distributed/shuffle/tests/test_rechunk.py
Comment (hendrikmakait) [inline distributed/shuffle/tests/test_rechunk.py:210]: XREF: #7816

File: distributed/reschedule.py
Comment (hendrikmakait) [inline distributed/reschedule.py:4]: I had to move this to avoid errors on import.

File: distributed/shuffle/tests/test_shuffle.py
Comment (fjetter) [inline distributed/shuffle/tests/test_shuffle.py:921]: If this is now able to recover we should add a couple of additional asserts

1. Assert that there is indeed a shuffle task on the dead worker
2. Assert that the output result is what is expected

Comment (fjetter) [reply distributed/shuffle/tests/test_shuffle.py:921]: I'm surprised that this change allows us to rerun. Do you understand why this works? You also reduced the data / number of partitions which makes me nervous in thinking that the now-dead worker just didn't run any shuffle tasks

**Review by fjetter — COMMENTED** [review]
Implementation looks good. There are a couple of questions around testing we should address before merging

File: distributed/shuffle/_scheduler_extension.py
Comment (hendrikmakait) [reply distributed/shuffle/_scheduler_extension.py:224]: I've added a docstring.

File: distributed/shuffle/tests/utils.py
Comment (hendrikmakait) [reply distributed/shuffle/tests/utils.py:93]: Generally, I'm also not a huge fan of this, but I think it is the most defensive way of testing that we do not rely on annotations. Let me disentangle your statement.

> This is testing properties of the current implementation that I consider mostly accidental / nice to haves but not hard requirements.

This is true, but it is also the implementation of annotations that is broken. If it were not broken, we would not need to fix this.

> For instance, I think it is a fair assumption to say that either all tasks or no tasks are annotated. This is generally how fusion and our graph manipulation works. The fact that the implementation can handle a mixture is nice but not required.

At the moment, I do not feel confident to make this claim. I suppose it's true, but I don't know what's broken with annotations at the moment. Also, this feels like an implementation-specific assumption. I could see fusion dropping annotations just for fused tasks if one does not pay attention. I would like to not have to test this, but here we are.

> I think it's also reasonably to say that annotations might be read during runtime and not just initially. I don't know when or how this is useful but I don't see why we should restrict ourselves to the current behavior.

I don't see where we restrict ourselves here. We are stripping the barrier and its dependents of annotations once they arrive on the scheduler's state machine. This makes no assumption about when they will be read. The only assumption this makes is that all the ways that annotations are broken will strip them before the tasks arrive at the scheduler. I think this is a fair assumption.

> I wouldn't want future development to be slowed down if they break these properties / because these tests fail.

Agreed.

File: distributed/shuffle/_scheduler_extension.py
Comment (j-bennet) [inline distributed/shuffle/_scheduler_extension.py:285]: Dummy question, sorry. The problem that you're fixing is tasks occasionally losing annotations. AFAIK this is applicable to all annotations, not just `shuffle` annotations, and your `ShuffleAnnotationChaosPlugin` is only killing`shuffle` annotation. Is there a possibility that in a real-world scenario, this new `shuffle_original_restrictions` will be lost too?

**Review by j-bennet — COMMENTED** [review]
I don't know enough about annotations and what they are used for, but simpler workflow is almost always better, and this is looks simpler.

I would be interested to see if this extra scheduling work (scheduling a task to a worker which may not be the right worker, and having to reschedule) leads to any slowdowns in realistically large workflows.

File: distributed/reschedule.py
Comment (hendrikmakait) [reply distributed/reschedule.py:4]: `distributed.exceptions` is a good suggestion, I've renamed the module. `Reschedule` has been around for a while, so I'd rather not rename it as this would be a breaking change (and completely unrelated to this PR).

File: distributed/shuffle/_scheduler_extension.py
Comment (hendrikmakait) [reply distributed/shuffle/_scheduler_extension.py:246]: Unless I have missed something, the `continue` would cause the loop to carry on with the next value and skip those lines.

File: distributed/shuffle/_scheduler_extension.py
Comment (hendrikmakait) [reply distributed/shuffle/_scheduler_extension.py:285]: Based on what we've seen, the current assumption is that `annotations` get lost _before_ the tasks make it to the scheduler.

File: distributed/shuffle/_scheduler_extension.py
Comment (j-bennet) [reply distributed/shuffle/_scheduler_extension.py:246]: No, you're good, I think I missed the `continue`. 👍

File: distributed/reschedule.py
Comment (j-bennet) [reply distributed/reschedule.py:4]: 👍

File: distributed/shuffle/tests/test_shuffle.py
Comment (fjetter) [reply distributed/shuffle/tests/test_shuffle.py]: thanks for digging in

File: distributed/shuffle/tests/utils.py
Comment (fjetter) [reply distributed/shuffle/tests/utils.py:93]: > At the moment, I do not feel confident to make this claim

Fair enough. It's true that it's hard to extrapolate what happens in the future. 

> I could see fusion dropping annotations just for fused tasks if one does not pay attention.

Indeed. My point is rather that if fusion happens this should affect all output keys. However, this is also a generalization that may not be true. 

> > I think it's also reasonably to say that annotations might be read during runtime and not just initially. I don't know when or how this is useful but I don't see why we should restrict ourselves to the current behavior.

> I don't see where we restrict ourselves here.

Sorry, I see how my statement is not very clear. This thing mutates annotations as soon as the barrier is transitioned to a specific state. If we read out the annotations before this already, this modification will not have *any* effect.
If I'm not mistaken, this is also what's happening in the current implementation. Once the first transfer task is running we're evaluating annotations and are fixing the mappings. This transition hook is only executed later, i.e. the mutated annotations will not have an effect on any of this.
If we later changed the behavior such that the annotations are evaluated later or multiple times we are suddenly hitting this test code which may no longer make sense.

This is all a bit academic. I will not block merging because of this.

> > I wouldn't want future development to be slowed down if they break these properties / because these tests fail.
> 
> Agreed.

What I meant saying is that red tests often discourage people to change something and once this becomes legacy code, the new developer generation may not fully understand the context about whether this is required or just nice-to-have behavior.
I raised my concerns. If future developers run into this, they can follow the breadcrumbs to this thread =)

**Review by fjetter — APPROVED** [review]
Good to go once CI is green-ish

**Comment by hendrikmakait** [comment]
It's not exactly green, but test failures appear to be unrelated, so I'll merge this in.

**Comment by mrocklin** [comment]
@rjzamora my guess is that p2p shuffling should be doable in dask-expr now.

**Comment by rjzamora** [comment]
>my guess is that p2p shuffling should be doable in dask-expr now.

Nice! Thanks for this @hendrikmakait !

**Pull request merged**
